server/src/test/java/org/elasticsearch/cluster/routing/allocation/allocator/BalancedShardsAllocatorTests.java

```diff
@@ -1104,6 +1106,160 @@ public void testShardMovementPriorityComparator() {
     }
 
+    public void testAssigmentPreferenceForUnassignedShards() {
+        final var notPreferredDecider = new AllocationDecider() {
+            @Override
+            public Decision canAllocate(ShardRouting shardRouting, RoutingNode node, RoutingAllocation allocation) {
+                final var nodeId = node.node().getId();
+                if (nodeId.startsWith("not-preferred")) {
+                    return Decision.NOT_PREFERRED;
+                } else if (nodeId.startsWith("yes")) {
+                    return Decision.YES;
+                } else if (nodeId.startsWith("no")) {
+                    return Decision.NO;
+                } else if (nodeId.startsWith("throttle")) {
+                    return Decision.THROTTLE;
+                } else {
+                    throw new AssertionError("unexpected node name: " + node.node().getName());
+                }
+            }
+        };
+
+        final var allocationService = new MockAllocationService(
+            new AllocationDeciders(List.of(notPreferredDecider)),
+            new TestGatewayAllocator(),
+            new BalancedShardsAllocator(BalancerSettings.DEFAULT, TEST_WRITE_LOAD_FORECASTER, new NodeNameDrivenBalancingWeightsFactory()),
+            () -> ClusterInfo.EMPTY,
+            SNAPSHOT_INFO_SERVICE_WITH_NO_SHARD_SIZES
+        );
+
+        // No allocation when NO
+        assertAssignedTo(allocationService, null, shuffledList("no"));
+        // NOT_PREFERRED over NO
+        assertAssignedTo(allocationService, "not-preferred", shuffledList("not-preferred", "no"));
+        // THROTTLE over NOT_PREFERRED/NO
+        assertAssignedTo(allocationService, null, shuffledList("throttle", "not-preferred", "no"));
+        // THROTTLE over NOT_PREFERRED
+        assertAssignedTo(allocationService, null, shuffledList("throttle", "not-preferred"));
+        // YES over THROTTLE/NO/NOT_PREFERRED
+        assertAssignedTo(allocationService, "yes", shuffledList("not-preferred", "yes", "throttle", "no"));
+        // prioritize YES/THROTTLE by weight
+        assertAssignedTo(allocationService, null, shuffledList("throttle-low", "yes-high", "yes"));
+        assertAssignedTo(allocationService, "yes-low", shuffledList("yes-low", "throttle-high"));
+        // prioritize YES over THROTTLE when weights equal
+        assertAssignedTo(allocationService, "yes-low", shuffledList("yes-low", "throttle-low"));
+        // prioritize YES by weight
+        assertAssignedTo(allocationService, "yes-low", shuffledList("yes-low", "yes", "yes-high"));
+        // prioritize NOT_PREFERRED by weight
+        assertAssignedTo(allocationService, "not-preferred-low", shuffledList("not-preferred-low", "not-preferred", "not-preferred-high"));
+    }
+
+    private List<String> shuffledList(String... strings) {
+        final List<String> list = new ArrayList<>();
+        Collections.addAll(list, strings);
+        Collections.shuffle(list, random());
+        return list;
+    }
```

Comment:
Nit: there is an existing `shuffledList(List<T>)` method that could possibly replace this.